huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go

```diff
@@ -0,0 +1,451 @@
+	v5 "github.com/huaweicloud/huaweicloud-sdk-go-v3/services/iotda/v5"
+	"github.com/huaweicloud/huaweicloud-sdk-go-v3/services/iotda/v5/model"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+const (
+	deviceStatusFrozen = "FROZEN"
+)
+
+func ResourceDevice() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: ResourceDeviceCreate,
+		UpdateContext: ResourceDeviceUpdate,
+		DeleteContext: ResourceDeviceDelete,
+		ReadContext:   ResourceDeviceRead,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(4, 256),
+					validation.StringMatch(regexp.MustCompile(stringRegxp), stringFormatMsg),
+				),
+			},
+
+			"node_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(4, 256),
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z-_0-9]*$`),
+						"Only letters, digits, underscores (_) and hyphens (-) are allowed."),
+				),
+			},
+
+			"space": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"product_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"device_id": { //keep same with console
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(4, 256),
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z-_0-9]*$`),
+						"Only letters, digits, underscores (_) and hyphens (-) are allowed."),
+				),
+			},
+
+			"secret": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				Sensitive:     true,
+				ConflictsWith: []string{"fingerprint"},
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(8, 32),
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z-_0-9]*$`),
+						"Only letters, digits, underscores (_) and hyphens (-) are allowed."),
+				),
+			},
+
+			"fingerprint": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"secret"},
+				ValidateFunc: func(i interface{}, k string) (warnings []string, errors []error) {
+					v, ok := i.(string)
+					if !ok {
+						errors = append(errors, fmt.Errorf("expected type of %s to be string", k))
+						return warnings, errors
+					}
+
+					if len(v) == 40 || len(v) == 64 {
```

Comment:
should be `if len(v) != 40 && len(v) != 64`?

Reply:
fix